src/concrete/ml/torch/hybrid_model.py

```diff
@@ -91,30 +191,97 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
         Returns:
             (torch.Tensor): The output tensor.
+
+        Raises:
+            ValueError: if fhe_mode is not supported
         """
-        if self.fhe_local_mode != "disable":
-            # for mypy
-            assert self.private_module is not None
+        # - disable: torch module
+        # - remote: client-server
+        # - simulate: compiled simulation
+        # - calibrate: calibration
+
+        if self.fhe_local_mode not in {FHEMode.DISABLE, FHEMode.CALIBRATE, FHEMode.REMOTE, None}:
+            # Using quantized module
             assert self.private_q_module is not None
-            y = self.private_q_module.forward(x.detach().numpy(), fhe=self.fhe_local_mode)
-            y = torch.Tensor(y)
-        elif self.private_module is not None:
-            if isinstance(x, torch.Tensor):
-                self.calibration_data.append(x.detach())
+            y = torch.Tensor(
+                self.private_q_module.forward(x.detach().numpy(), fhe=self.fhe_local_mode.value)
+            )
+        elif self.fhe_local_mode == FHEMode.DISABLE:
+            # Calling torch
+            assert self.private_module is not None
+            y = self.private_module.forward(
+                x.detach(),
+            )
+            assert isinstance(y, torch.Tensor)
+        elif self.fhe_local_mode == FHEMode.CALIBRATE:
+            # Calling torch + gathering calibration data
+            assert self.private_module is not None
+            self.calibration_data.append(x.detach())
             y = self.private_module(x)
+            assert isinstance(y, torch.Tensor)
         # TODO: https://github.com/zama-ai/concrete-ml-internal/issues/3869
-        # else:
-        #     y = self.remote_call(x)
+        elif self.fhe_local_mode == FHEMode.REMOTE:  # pragma:no cover
+            # Remote call
+            y = self.remote_call(x)
+        else:  # pragma:no cover
+            # Shouldn't happen
+            raise ValueError(f"{self.fhe_local_mode} is not recognized")
         return y
 
-    def remote_call(self, x: torch.Tensor):
+    def remote_call(self, x: torch.Tensor) -> torch.Tensor:  # pragma:no cover
         """Call the remote server to get the private module inference.
 
         Args:
             x (torch.Tensor): The input tensor.
+
+        Returns:
+            torch.Tensor: The result of the FHE computation
         """
         # TODO: https://github.com/zama-ai/concrete-ml-internal/issues/3869
         # implement server call and client initialization
+        base_device = x.device
```

Comment:
I think the following needs to be documented a bit, not so sure what's going on in some parts

Reply:
I'll add inline comments

Reply:
thanks, looks better !